Merge Parallel Trace Abstraction into Dev

MaxBarth95 opened the pull request:
Updated

I prepared this branch to merge my parallel trace abstraction approach into main.
Feel free to take a look already, make suggestions, and ideally tell me how you want it to be refactored.
Unfortunately my changes  affects quite a few classes.

Before merging, I need to run a few more experiments to ensure there are no unwanted side effects. 

The following changes have been made:

-  ParallelNwaCegarLoop.java
Contains the main CEGAR loop

- CegarNwaWorkerThread.java
Is the corresponding class that does the trace check, interpolation, and generalization of the resulting automaton

- Some changes are made to the refinement strategy stuff. In the CegarNwaWorkerThread.java I can for example
(se commented code) decide to do LoopAcceleration after seeing a PathProgram 7 times.

 - IsEmptyParallel
The search now searches for a Prefix of a counterexamples that is different from all counterexamples in a Set given as parameter.  Then IsEmptyHeuristic is called with waypoints. Those waypoints are the Prefix and the IsEmptyHeuristic search follows those waypoints and after the last waypoints it searches for an accepting state.
It will not backtrack beyond the last waypoint. 

- TransferBetweenMainAndWorker
This is now the more elegant solution to our script problem. Each worker has one such transfer class and it manages all transfers in both direction. It transfers the CfgSmtToolkit, Automata, Counterexamples. To be precise, transformulas, IProgramVars everything that the CFGSmtToolkit touches

- Wintesses
To create a violation witness i need the sort of a value in the TypeSortTranslator. This sorts come from worker smtcfgtoolkit scripts. Since the TypeSortTranslator is created at the very start when translating to Boogie. I need a way to tell the TypeSortTranslator what the script is, that created this sort. I do this by passing it through the programexecution

MaxBarth95 committed:
Test Generation: Added support for ONE_CEGAR_PER_ERROR_LOCATION and Timelimit

MaxBarth95 committed:
Reuse when Unsat solang kein function call im Trace ist

MaxBarth95 committed:
Nur reuse, wenn keine function calls in trace

MaxBarth95 committed:
Normal Reuse but Reuse when unsat only when no functions calls in trace

MaxBarth95 committed:
Only Reuse when Unsat if CurrentVA not in function call other than main

MaxBarth95 committed:
Second try, reuse only when not in function and reuse unsat only when not in function

MaxBarth95 committed:
Reactivated VAOrder, use last checked for reuse

MaxBarth95 committed:
Fix reuseunsat, readde no reuse if negated

MaxBarth95 committed:
Normal Reuse, only reuse vas annotated to test-goals that are not in a function

MaxBarth95 committed:
minor fix

MaxBarth95 committed:
Removed redundant SatCheck

MaxBarth95 committed:
Possible Fix for having error automaton but not an active test-goal

MaxBarth95 committed:
BvToInt: Fix Unsoundness in SMT-Comp23

MaxBarth95 committed:
Instead of removing covered test goals from the A* goal set, we add them to the error automaton

MaxBarth95 committed:
Make ReuseUnsat dependend on procedures calling location

MaxBarth95 committed:
Fixed Default Reuse

MaxBarth95 committed:
Fix Match Prefix WIP

MaxBarth95 committed:
Match prefix for resuse Fix

MaxBarth95 committed:
minor fix

MaxBarth95 committed:
Removed prints

MaxBarth95 committed:
Version where we reuse always and only use optimization if prefixes match

MaxBarth95 committed:
Added setting for calloc

MaxBarth95 committed:
Added support for Reuse + Mutligoal needs evaluation

MaxBarth95 committed:
Removed Concrete execution method for now

MaxBarth95 committed:
Prepared STTT Evaluation:

New Statistics
Reuse is tried on the CFG script
Removed old statistics

MaxBarth95 committed:
Bug Fix, unlock script if error occurs

MaxBarth95 committed:
Bugfix unknown var in reusescript

MaxBarth95 committed:
Fix if value is real but sort is int

MaxBarth95 committed:
assert to cfg script even if no reuse

MaxBarth95 committed:
Reveted using cfg script for reuse, overhead too big

MaxBarth95 committed:
fix forgott to pop() and provide no programExecution in case of reuse

MaxBarth95 committed:
Fixed that we reuse for the first goal again

MaxBarth95 committed:
Switch Floats and Double. Why was it the other way around?

MaxBarth95 committed:
Test Export: Potential fix for Floats

MaxBarth95 committed:
Parallel Trace Abstraction v1

MaxBarth95 committed:
Missing files

MaxBarth95 committed:
Fix Build

MaxBarth95 committed:
No more busy waiting.

Addes support for test-case generation.
- error autoamta
- no support for reuse yet

danieldietsch committed:
pin jdk and maven versions

(cherry picked from commit 9a9590fe828969759f7a0439c8f5f96b5d7840ec)

cherry pick java version

MaxBarth95 committed:
Added support for parallel test case generation.

Each thread analyses a different test goal.

If there is an available thread but not available test goal, we reduce to one thread.

MaxBarth95 committed:
changed logger back, leads to bugs in Ultiamte.py

MaxBarth95 committed:
Terminate if thread throws exception
refactored a bit, moved termtransferrer calls to HistoryRecordingScript
added Termvariable mapping to HistoryRecordingScript
This fixed some bugs benchmark: Ackermann02.c

MaxBarth95 committed:
cleaned up prints

MaxBarth95 committed:
small rework of metadata file extraction
strategy default is now the user provided strategy

MaxBarth95 committed:
Added Setting to switch between parallel and seq

MaxBarth95 committed:
TODO removed erroratuomaton statistics for evaluation. Readd Later!

MaxBarth95 committed:
Added setting for maxThreads

MaxBarth95 committed:
minor fix

MaxBarth95 committed:
another minor fix

MaxBarth95 committed:
deactivate gc after cegar iteration
now we do difference in worker and main, not sure if this is what matthias thought of

MaxBarth95 committed:
if we log with thid we cannot determine ultimate verison

MaxBarth95 committed:
Some advances / bugfixes for parallel in the incrHoareTripleChecker

MaxBarth95 committed:
maxThread is now available cores -1
Fixed bugs in incr. hoaretriplechecker
Fxied bugs related to Transformulas

MaxBarth95 committed:
Added Search Strategy for Paralle search

MaxBarth95 committed:
Some Bug fixes
Better exception print
search doesnt mark as visited as long as counterexamples tracking is not done

MaxBarth95 committed:
reworked the search,

Now we track all previous counterexamples may cost too much memory but its for debugging
we use shutdownNow instead of shutdown

MaxBarth95 committed:
remembering all counterexamples takes too much memory

MaxBarth95 committed:
Changed Search, and improved the loop in case no counterexample can be found

MaxBarth95 committed:
Dont save all counterexamples

MaxBarth95 committed:
ReachSafety: works without minimization
Changed testgoal setting such that no more error automata are constructed
Leads to faster SAT tracechecks and less idle time.

MaxBarth95 committed:
If counterexample is in allcounterexamples then do not check.
Removed the assert

MaxBarth95 committed:
All worker get the inital abstraction and not an intermediate result

MaxBarth95 committed:
unsound in reach safety, say unsafe instead of safe
fixed search hopefully,
added minimizebastraction for reach safety, here we minimize once per worker who returns flyod hoare

MaxBarth95 committed:
Reset IsEmpty
Added IsEmptyParallel and i hope it works now
Added Setting for active/all cex
Added super.isAbstractionEmpty to termiante early if it is.

MaxBarth95 committed:
Added setting for minimization

MaxBarth95 committed:
fixed some isEmptyParallel bugs

warning atm testgoals are error locations in reach safety

MaxBarth95 committed:
Cant fix the IsEmptyParallel Bug atm, dont know where it comes from
IsEmptyParallel and IsEmpty both find only once successor along the run.
However, IsEmptyParallel finds only a path already taken before while IsEmpty seems to find a new one.
However, the check is based on the trace hash. Not sure how reliable that is

MaxBarth95 committed:
Deactivated debug asserts to for trial run

MaxBarth95 committed:
Looks like the bug is fixed:

Need to check the symbol not only the satate and we need to use isFinal
instead of goal set.

-> Debug run and then refaktore everything implemented for bugfixing

MaxBarth95 committed:
this wont work with use test goal set as goal set

MaxBarth95 committed:
Some minor refactoring

MaxBarth95 committed:
Deactivated counterexample report
Activated reach safety result report
traceCHeckUtils fixde a concurrency bug. Fix may cause issues
Still some problem in isparalle i think hard to reproduce
Fixed is parallel stateK at calls

MaxBarth95 committed:
reorganised the loop, now we wait if we didnt find a new cex for an abstraction refinement
we serach for a new cex after refinement and we refine asap

MaxBarth95 committed:
long description of unprovable and unsafe result doesnt work since the sort comes from a unknown script

MaxBarth95 committed:
terminte if isEmpty sais true even if threads are running
check after refinement if current cex became redundant
stats for that

MaxBarth95 committed:
missing file

MaxBarth95 committed:
fixed recursions
fixed naming hopefully
deal with error automata first

MaxBarth95 committed:
deactivate recusive fix

MaxBarth95 committed:
small change to loop

MaxBarth95 committed:
deactiavete worker cancel

MaxBarth95 committed:
was still writing test cases
fixed early termination if feasible

MaxBarth95 committed:
removed the cancelation completely

MaxBarth95 committed:
reversed change

MaxBarth95 committed:
refactored the loop

MaxBarth95 committed:
Merge branch 'dev' into TestGeneration

# Conflicts:
#	Jenkinsfile
#	Jenkinsfile.deploy_website
#	Jenkinsfile.nightly
#	releaseScripts/default/adds/Ultimate.py
#	trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CExpressionTranslator.java
#	trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/CHandler.java
#	trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/TranslationSettings.java
#	trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java
#	trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/preferences/CACSLPreferenceInitializer.java
#	trunk/source/CoreRCP/src/de/uni_freiburg/informatik/ultimate/core/coreplugin/services/ModelTranslationContainer.java
#	trunk/source/Library-BoogieAST/src/de/uni_freiburg/informatik/ultimate/boogie/ast/BoogieASTNode.java
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/transitions/TransFormulaBuilder.java
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/transitions/TransFormulaUtils.java
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/smt/predicates/PredicateUnifier.java
#	trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/smt/predicates/TermVarsFuns.java
#	trunk/source/Library-SmtLibUtils/src/de/uni_freiburg/informatik/ultimate/lib/smtlibutils/bvinttranslation/BvToIntTransferrer.java
#	trunk/source/Library-SmtLibUtils/src/de/uni_freiburg/informatik/ultimate/lib/smtlibutils/bvinttranslation/TranslationManager.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/singletracecheck/AnnotateAndAsserter.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/singletracecheck/InterpolatingTraceCheckCraig.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/singletracecheck/NestedInterpolantsBuilder.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/singletracecheck/TraceCheck.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/singletracecheck/TraceCheckSpWp.java
#	trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/singletracecheck/TraceCheckUtils.java
#	trunk/source/Library-UltimateCore/src/de/uni_freiburg/informatik/ultimate/core/lib/models/annotation/Check.java
#	trunk/source/Library-UltimateCore/src/de/uni_freiburg/informatik/ultimate/core/lib/results/UnprovableResult.java
#	trunk/source/Library-UltimateTest/src/de/uni_freiburg/informatik/ultimate/test/decider/SafetyCheckTestResultDecider.java
#	trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/preferences/RcfgPreferenceInitializer.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/AbstractCegarLoop.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/BasicCegarLoop.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarLoopFactory.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarLoopStatisticsDefinitions.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarLoopStatisticsGenerator.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/NwaCegarLoop.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TraceAbstractionStarter.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/preferences/TAPreferences.java
#	trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/tracehandling/StrategyFactory.java
#	trunk/source/WitnessPrinter/src/de/uni_freiburg/informatik/ultimate/witnessprinter/WitnessPrinter.java

MaxBarth95 committed:
fixing merge issues

MaxBarth95 committed:
fixing merge, deactivated reuse und annotate and assert and trace check

MaxBarth95 committed:
fixing merge

MaxBarth95 committed:
fixing merge

MaxBarth95 committed:
fixing merge

MaxBarth95 committed:
fixing merge, added dependencies for test case generation in trace check
fixed strategyfactory

MaxBarth95 committed:
fixing merge

MaxBarth95 committed:
Fixed some Parallel stuff lost in merge

MaxBarth95 committed:
always try bfs first

MaxBarth95 committed:
use counterexample class correclty to fix bug where we dont check for perfect interpolants

MaxBarth95 committed:
BugFix interpolation with unsat core produces invalid hoare triple

MaxBarth95 committed:
isEmptyParallel no longer throws an exception if it fails to construct the cex

MaxBarth95 committed:
added patchcache to parallel loop so that we do loop acceleration

MaxBarth95 committed:
deactivated test regex in ultaimte.py!
fixed build

MaxBarth95 committed:
less memory consumption in isemptyparallel
reset Ultiamte.py to default (No testcasegeneration)

MaxBarth95 committed:
isEmptyParallel deactivated the recursion fix again.
Maybe what fixed it was the assertion order

MaxBarth95 committed:
fixed some warnings.
Somehow we got slower, maybe this fixes it otherwise its the pathcache
reachtivated somer asserts

MaxBarth95 committed:
give worker a copy of patchprogramcache
I suspect otherwise we synchronize

MaxBarth95 committed:
Made some more copies of globals maybe useless

there is still a problem in isEmptyParallel,
deactivated the assert for now

MaxBarth95 committed:
IsEmptyParallel can become very expensive
Build in a limit for backtracks 30 atm

MaxBarth95 committed:
Added more statistics

MaxBarth95 committed:
check if programcache is the problem

MaxBarth95 committed:
making a real copy of programcache

MaxBarth95 committed:
some more statistics

MaxBarth95 committed:
start as many threads as possible
dfs if fail
timeout search 5 sec
catch exceptions
statistics

MaxBarth95 committed:
IsEmptyParallel: Fixed that we get 2 returns for the same call
by processing summaries first and dont process returns if we have a summary

UnproveableResult path not supported, sort unkown since it comes from workerscript

PathProgramCache fixed the order in which we copy and fill the cache

MaxBarth95 committed:
No witness printing in Parallel CEGAR

MaxBarth95 committed:
deactivated concurrency CEGAR  for evaluation

MaxBarth95 committed:
Tried sth. might introduce a datarace on mAbstraction
The worker now uses a getter to get the abstraction,
ideally saves memory and runtime on generalization

MaxBarth95 committed:
isEmtpyParallel stackOVerflow prevention

MaxBarth95 committed:
removed backtransfer in qelim

MaxBarth95 committed:
orginal program cache was the copy

MaxBarth95 committed:
added script exit to tracecheckSPWP

MaxBarth95 committed:
removed tcScript.exit()

MaxBarth95 committed:
Fixed Bug in termiantion

MaxBarth95 committed:
reactivate concurrency

MaxBarth95 committed:
testing new naming for variables that depends on trace hash.
Check if this gets rid of outliers in scaling

MaxBarth95 committed:
Reversed Naming.
Doesnt work like this we need sth far more complex
if it is even possible to have a naming independent of the current cegar iteration

MaxBarth95 committed:
trying to fix recursion unsoundness in search

MaxBarth95 committed:
Trying to fix Recursion

MaxBarth95 committed:
if search not correct return null instead of incorrect word

There is still a bug in isemptyparallel that provides us with counterexamples
not in the program. This case is captured by an is accrpted check

MaxBarth95 committed:
Started Major refactoring,
refactored IsEmptyParallel now inherits from IsEmpty
 and how it is called in ParallelCEGAR loop.

Addtionally added TraceCheckSPWP exit script, might be redundant

MaxBarth95 committed:
Refactored the Loop

MaxBarth95 committed:
trying new search that focues on circumventing loops

MaxBarth95 committed:
backup for parallel strategy, not yet working

MaxBarth95 committed:
Framework for parallel strategy is done:
TODO actual strategies
Default setting

MaxBarth95 committed:
Added Support for parallel loop acceleration
New version with parallel path program check

MaxBarth95 committed:
fixed more for parallel loop acceleration

MaxBarth95 committed:
BugFix in Parallel Strategy LoopAcceleration, only do LA once per PP. only enhance if 2 or more imperfect sequences, TODO create union of imperfect

MaxBarth95 committed:
1 thread per executor, 2 threads per  countereaxmple. Per cex we queue craig first then fp bp. then we try one time loop acceleration the second time we see the path program

MaxBarth95 committed:
Thread Limit is now limit of Groups (PathPRogram) times limit of strategies

MaxBarth95 committed:
New dynamic executor, executor has a limit of threadlimit / pathprograms seen

MaxBarth95 committed:
changed how running threads is counted, now its basically running path programs

MaxBarth95 committed:
Added Union for NWAs, removed totalization nad changed the resultCheck, it failed when both arguments of the union were not disjoint

MaxBarth95 committed:
missing files

MaxBarth95 committed:
Added a worker thread that can only generalize a given automaton

MaxBarth95 committed:
removed local benchmark

MaxBarth95 committed:
removed GeneralizationWorker, cannot be used at the moment. We fail when CegarWorkerThread does not generalize itself

MaxBarth95 committed:
added enum for different generalization modes in worker thread.
Cleaned the CegarWorkerThread class

MaxBarth95 committed:
Added a new strategy that does not interpolate at all

MaxBarth95 committed:
Preparing Merge, reset files without meaningfull change to dev version

MaxBarth95 committed:
resetting files used only for testgen to dev

MaxBarth95 committed:
Merge branch 'TestGeneration' of https://github.com/ultimate-pa/ultimate into TestGeneration

MaxBarth95 committed:
Adding new files from dev for libraries

MaxBarth95 committed:
resetting files to dev only used in testgen

MaxBarth95 committed:
Removed more code related to testgeneration

MaxBarth95 committed:
Merge branch 'dev' into TestGeneration

MaxBarth95 committed:
Fixed Merge,
Removed more unsued testgen stuff
Changed the setting for activating the parallel cegar loop

MaxBarth95 committed:
Cleanup of some style differences between branch and dev
cleanup of some code

MaxBarth95 committed:
IsEmpty used no loop search! changed back to default

Reworked the parallel strategy and
 how thread limit and running threads is calculated

MaxBarth95 committed:
Cleanup some code, removed unused stuff

MaxBarth95 committed:
Removed unused class

MaxBarth95 committed:
Minor changes

MaxBarth95 committed:
Added comment to warnings from VariableManager

MaxBarth95 committed:
Fixed ThreadPoolCalculation

MaxBarth95 committed:
Fixed Executor size calculation
Can use default strategies now, too

MaxBarth95 committed:
Trying to support witnesses, by making use of the HistoryRecordingScript termtransferrer

MaxBarth95 committed:
Added Parallel Strategy for Bit-Precise mode

MaxBarth95 committed:
Added new parallelization that applies multiple worker per cex and a setting for it

MaxBarth95 committed:
missing file

MaxBarth95 committed:
Added a new strategy / mode
At the start in our serach we visit loops only once,
hoping for more spred.

During that time we only apply Craig interpolation

Hoping to cover a large portion of the state space quick.

When we no longer find cex we allow multiple per loop.
 Applying different assertion orders

MaxBarth95 committed:
Removed DFS, after failed serach

after swtiching mode, search again

MaxBarth95 committed:
Merge, removed some more testing stuff from abstractcegarloop

MaxBarth95 committed:
deactivate loop acceleration

MaxBarth95 committed:
Removed Futures, when using a blocking queue,
 using futures too leads to sleeping twice

Removed unused setting

Fixed Bug in TraceCHeck affecting NWA

MaxBarth95 committed:
Renamed ParallelCEGARLoop to ParallelNWACegarLoop and the worker too

MaxBarth95 committed:
renaming

remove map after its no longer needed

MaxBarth95 committed:
Renamed parallel cegar loop

schuessf commented on trunk/source/Library-IcfgTransformer/src/de/uni_freiburg/informatik/ultimate/icfgtransformer/loopacceleration/jordan/JordanLoopAcceleration.java:
```diff
@@ -304,14 +312,17 @@ private static ClosedFormOfUpdate computeClosedFormOfUpdate(final ManagedScript
 		for (final IProgramVar pv : su.getReadonlyVars()) {
-			final Term oldValue = substitutionMapping.put(pv.getTermVariable(), inVars.get(pv));
+			final Term oldValue =
+					substitutionMapping.put((TermVariable) ((HistoryRecordingScript) mgdScript.getScript())
```
I think we should look for a better solution here. We should not cast every script to `HistoryRecordingScript` to access the method `transferTermToWorker`.

Also, where is it necessary to transfer the terms? It looks like many locations are affected, maybe there is also a more elegant solution for that.

MaxBarth95 replied:
Yes a better solution would be great, so far i couldnt think of any.

We basically need it everywhere:
- Its necessary before we checksat or assert
- Before we build a formula we have to make sure the components are from the same script
- When we get a termVar representing a IProgramVar 
- When we read a transformula, In, Out vars etc.
- When we create fresh predicates or use an old one

schuessf commented on trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/transitions/UnmodifiableTransFormula.java:
```diff
@@ -227,8 +270,11 @@ private boolean doConstantConsistencyCheck() {
-			consistent &= constantsInFormula.contains(programConsts.getDefaultConstant());
+			nonTheoryConstantTerms.add((ApplicationTerm) ((HistoryRecordingScript) mMgdScript.getScript())
+					.transferTermToWorker(programConsts.getDefaultConstant()));
+			// assert mFormula.getTheory().equals(programConsts.getDefaultConstant().getTheory());
```
Please delete outcommented code

MaxBarth95 replied:
deleted

schuessf commented on trunk/source/Library-TraceCheckerUtils/META-INF/MANIFEST.MF:
```diff
@@ -25,3 +25,6 @@ Require-Bundle: de.uni_freiburg.informatik.ultimate.lib.core,
  de.uni_freiburg.informatik.ultimate.smtinterpol,
  de.uni_freiburg.informatik.ultimate.smtsolver.external
 Automatic-Module-Name: de.uni.freiburg.informatik.ultimate.lib.tracecheckerutils
+Import-Package: de.uni_freiburg.informatik.ultimate.boogie.ast,
```
I guess, we should not use any dependency to Boogie in this plugin. Why did you add it?

MaxBarth95 replied:
deleted

schuessf commented on trunk/source/Library-TraceCheckerUtils/META-INF/MANIFEST.MF:
```diff
@@ -25,3 +25,6 @@ Require-Bundle: de.uni_freiburg.informatik.ultimate.lib.core,
 Automatic-Module-Name: de.uni.freiburg.informatik.ultimate.lib.tracecheckerutils
+Import-Package: de.uni_freiburg.informatik.ultimate.boogie.ast,
+ de.uni_freiburg.informatik.ultimate.plugins.generator.rcfgbuilder.cfg,
+ de.uni_freiburg.informatik.ultimate.plugins.generator.rcfgbuilder.preferences
```
Why do we need settings of the `RCFGBuilder` here? This should be avoided.

MaxBarth95 replied:
deleted

schuessf commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/preferences/TraceAbstractionPreferenceInitializer.java:
```diff
@@ -1066,4 +1111,29 @@ public enum OrderOfErrorLocations {
+	 *
+	 * @author Max Barth (Max.Barth@gmx.de)
+	 */
+	public enum TestGenerationMode {
```
This has nothing to do with parallel trace abstraction, right?
Also, is this enum used anywhere?

MaxBarth95 replied:
deleted

schuessf reviewed (commented):
Thanks for your work! Running trace abstraction in parallel can be really useful to reduce the walltime in practice (where we don't care about the CPU time).
I did not yet look into this PR in detail, but I have some more general comments.
I can see that it is complicated to manage scripts for different threads, but the current solution does not seem nice (it introduces some possibly unsafe casts) and complicates much of the existing code. Therefore, we should rather look for a more elegant solution before merging this PR.

MaxBarth95 committed:
Removed remains from testgen

MaxBarth95 committed:
Removed commented code

MaxBarth95 committed:
Removed leftover from Testgen

MaxBarth95 committed:
All of them are not needed

maul-esel commented:
Thanks for your work on this. Bringing parallelization to Ultimate is a big step, and it will be great to get it merged back into `dev`.

I think what we need here is an understanding for the architecture of your concurrent analysis, so that we can identify suitable interfaces between the workers and the main analysis thread. Then, the code can be refactored to handle all the business of this interaction -- primarily, it seems, the transferring of terms between scripts -- at well-defined points of these interfaces; such that separation of concerns is maintained.

We already have some components in Ultimate using different scripts, perhaps they could serve as inspiration:
- Trace checks use their own scripts. They must transfer transition formulas from the CFG script to the trace check script, and interpolants / counterexamples from the trace check script to the CFG script.
- GemCutter's independence checks use their own script. Here, predicates and transition formulas must be transferred to the script, and (since recently) commutativity condition predicates must be transferred back.

The current approach -- transferring these formulas in the most diverse set of classes, from loop acceleration over TermVarsFuns to SSA building code -- is extremely brittle. Who is to say that there is not some code path that is triggered only by few benchmarks, where the transferral is still missing? Any future change to Ultimate, e.g. some new kind of trace check, might introduce such cases. etc. I think we need a more robust solution.

MaxBarth95 committed:
Reactivated long description for results with program execution.

MaxBarth95 committed:
minor style change

MaxBarth95 commented:
Sounds like a plan. You dont have to convince me that this needs some refactoring :) 
But before I do a refactoring i want us to agree on a suitable solution.

One thing we can do based on a suggestion of Matthias:
- Instead of using the HistoryRecordingScript we introduce a new Wrapper script.
This checks for every term and sort it sees, if its from its own theory and if not, it transfers from Main to Worker.
This will cleanup some clases, I'm guessing 30%.

- Additionally, we would need a new Substitution that transfers before its applies the substitution.

- Then we we are left with with the cases, where we have IProgramVars and their TermVariables and do renaming if I remember correctly

MaxBarth95 committed:
Added serach strategy that does  not allow loops at all,
Adapted Parallel refinement strategy,
now scales down after all assertion orders have been tried

MaxBarth95 committed:
Fixed some Bugs

IsEmptyParallel: We have to track the calls,
I readded it, this might cause problems lets see,
but without we cannot find cex in simple cases

maul-esel commented:
That may be a possibility, but it introduces somewhat magic behaviour ;-) And you say it doesn't cover all cases.

Is there a reason against simply transferring every object at the point where the object is handed from the main thread to a worker, or from a worker to the main thread?

MaxBarth95 committed:
Changing the amount of Running threads when updating executor sizes does not work atm

schuessf commented:
I was also thinking about something similar:
Would it work in priniciple to create a copy of the entire CFG, but with a fresh script for each worker? Then the the main thread could just do that and transfer the terms returned from the worker to the main script again.
Always creating an entirely new CFG would be really inefficient of course, but maybe we could just create a CFG implementation that wraps an exisiting CFG and transfers the terms on demand (and caches them).
If this does not work, you can tell me why and this way we might get towards a nicer solution 😉

MaxBarth95 committed:
reworked the search now uses a prefix that isEmptyHeuristic follows before exploring new states, fixed unknown result in worker  trace check

MaxBarth95 committed:
Added No Loop mode to IsEmptyHeuristic

MaxBarth95 committed:
moving towards continues worker only and transfer of automata. taken form testgen. removed a lot of stuff

MaxBarth95 committed:
removed parallel refinement strategy

MaxBarth95 committed:
removed parallel strategy

MaxBarth95 committed:
made StmtSequence, Call and Return public

MaxBarth95 committed:
give worker script through to TypeSortTranslator for backtranslation to C

MaxBarth95 committed:
Rebased all files with the old HistoryRecordingScript termtransfer

MaxBarth95 committed:
removed old file

MaxBarth95 committed:
rebased MonitoredProcess

MaxBarth95 committed:
reverted change, not sure what it does

MaxBarth95 committed:
Merge branch 'dev' into merge/mb/parallel-trace-abstraction

MaxBarth95 committed:
fix build

MaxBarth95 committed:
2 Bug fixes

MaxBarth95 committed:
added setPayload such that we can copy the payload to the worker edges

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmpty.java:
```diff
@@ -285,12 +289,41 @@ private IsEmpty(final AutomataLibraryServices services,
+			mLogger.info(exitMessage());
+		}
+	}
+
```
It seems this constructor is no longer needed? The only extra parameter is unused.

maul-esel replied:
Ok, now the extra parameter _is_ used. Nevertheless, please avoid duplicating all the other constructor code. I.e., just make the old constructor call this constructor with `initializeOnly = false`.

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyHeuristic.java:
```diff
@@ -83,7 +84,9 @@ public final class IsEmptyHeuristic<LETTER, STATE> extends UnaryNwaOperation<LET
-
+	protected final List<Pair<STATE, LETTER>> mWayPoints;
+	boolean stillFollowingWayPoint = true;
+	boolean mDontVisitLoops = false;
```
- Please indicate visibility for these fields. Is there any reason to not make them private?
- The first one also does not conform to our naming conventions.

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyHeuristic.java:
```diff
@@ -64,6 +64,7 @@
 import de.uni_freiburg.informatik.ultimate.util.CoreUtil;
 import de.uni_freiburg.informatik.ultimate.util.HashUtils;
 import de.uni_freiburg.informatik.ultimate.util.datastructures.HashedPriorityQueue;
+import de.uni_freiburg.informatik.ultimate.util.datastructures.relation.Pair;
```
Are the changes in this class related to parallel TAR? The waypoint stuff seems witness-related?

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyHeuristic.java:
```diff
@@ -1133,6 +1205,30 @@ static <STATE, LETTER> IHeuristic<STATE, LETTER> getHeuristic(final AStarHeurist
+		 * @param <LETTER>
+		 * @return
+		 */
+		static <STATE, LETTER> IHeuristic<STATE, LETTER> getParallelHeuristic() {
```
Please document the meaning of this heuristic.

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyParallel.java:
```diff
@@ -0,0 +1,861 @@
+ * Non-terminating if every existing counterexample is in the set but the state space is infinite.
+ *
+ * TODO: Has issues with recursive function calls, sometimes finds path that doesnt exist. -> for now we check
+ * isAccepted after the search to find these cases
```
Has this bug been fixed?

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyParallel.java:
Has this been fixed?

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyParallel.java:
please remove commented-out code

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyParallel.java:
should these fields be private final?

maul-esel replied:
no?

maul-esel commented on trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/boogie/TypeSortTranslator.java:
```diff
@@ -121,7 +124,11 @@ public TypeSortTranslator(final Script script, final IUltimateServiceProvider se
 
-	public IBoogieType getType(final Sort sort) {
+	public IBoogieType getType(Sort sort) {
+		if(mCfgScriptFromWorker != null) {
```
Please run auto-formatting on all modified projects.

maul-esel commented on trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/boogie/TypeSortTranslator.java:
```diff
@@ -273,4 +280,8 @@ private void cacheSort(final IBoogieType boogieType, final Sort result) {
 	}
 
+	public void setCfgScriptFromWorker(ManagedScript cfgScriptFromWorker) {
+		mCfgScriptFromWorker = cfgScriptFromWorker;
```
Who calls this? Is it necessary to make the class mutable?

maul-esel replied:
Transferring between scripts does not seem like a primary task for `TypeSortTranslator`. Could `RCFGBacktranslator` not do the transfer itself, before calling `getType()` ?

MaxBarth95 replied:
RCFGBacktranslator is definetly the wrong place to do this. It doesnt even have a script.
I moved it now to Term2Expression, even tho i found TypeSortTranslator very fitting.

maul-esel commented on trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/IcfgProgramExecution.java:
```diff
@@ -359,5 +361,12 @@ public IBacktranslationValueProvider<L, Term> getBacktranslationValueProvider()
 	}
+	
+	public void setOriginCfgScript(ManagedScript managedScript) {
+		mScript = managedScript;
```
Is it necessary to make this class mutable?

maul-esel replied:
If `mScript` is used "on creation", as your comment above states, could you not set this in the constructor? (and make it a final field?)

Or, alternatively, pass the script alongside (but not as part of) the execution?

maul-esel commented on trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/tracehandling/AutomatonFreeRefinementEngine.java:
```diff
@@ -76,6 +76,19 @@ public final class AutomatonFreeRefinementEngine<L extends IAction>
+		mFeasibility = executeStrategy();
+		mRefinementEngineStatistics.finishRefinementEngineRun();
+	}
+
```
This duplicates code from the constructor below. Call `this(...)` instead.

maul-esel commented on trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/tracehandling/AutomatonFreeRefinementEngine.java:
```diff
@@ -126,7 +139,7 @@ private LBool executeStrategy() {
 
 		// trace was feasible, return
-		if (feasibilityResult == LBool.SAT) {
+		if (feasibilityResult == LBool.SAT || mQuickCheck) {
```
This seems wrong -- the log should only be written if the trace is truly feasible.

maul-esel commented on trunk/source/Library-SmtLibUtils/src/de/uni_freiburg/informatik/ultimate/lib/smtlibutils/ManagedScript.java:
```diff
@@ -144,6 +144,10 @@ public LBool assertTerm(final Object lockOwner, final Term term) throws SMTLIBEx
+		if (Thread.interrupted()) {
+			mScript.exit();
+			throw new SMTLIBException("Thread Was Interrupted, crashing before checkSat");
+		}
```
What is the purpose of this code? Why do you check this only here? Would it be possible to use Ultimate's internal timeout mechanism instead?

maul-esel replied:
ping

maul-esel commented on trunk/source/Library-SmtLibUtils/src/de/uni_freiburg/informatik/ultimate/lib/smtlibutils/ManagedScript.java:
```diff
@@ -327,9 +331,11 @@ public TermVariable constructFreshTermVariable(final String name, final Sort sor
 		public TermVariable constructFreshCopy(final TermVariable tv) {
 			String basename = mTv2Basename.get(tv);
 			if (basename == null) {
+				// Variable Names are not synchronized in Parallel CEGAR loop, this warning will be triggered!
```
Is this an issue that could be fixed?

maul-esel replied:
ping

maul-esel replied:
Often such warnings lead to bugs later on, or make debugging extremely difficult (because there are distinct variables that have the same name)

maul-esel commented on trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/singletracecheck/InterpolatingTraceCheckCraig.java:
```diff
@@ -119,6 +119,7 @@ public InterpolatingTraceCheckCraig(final IPredicate precondition, final IPredic
 			mInterpolantComputationStatus =
 					new InterpolantComputationStatus(ItpErrorStatus.SMT_SOLVER_CANNOT_INTERPOLATE_INPUT, null);
 		}
+		mgdScriptTc.getScript().exit();
```
I see you added this to several trace checks. What is the reason behind this?

maul-esel replied:
ping

maul-esel commented on trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/Call.java:
```diff
@@ -86,5 +87,8 @@ public UnmodifiableTransFormula getLocalVarsAssignment() {
+	
+	public void setPayload(IPayload payload) {
+		mPayload = payload;
+	}
```
Is it necessary to make the payload mutable?

maul-esel replied:
ping

maul-esel commented on trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/Call.java:
```diff
@@ -52,7 +53,7 @@ public class Call extends CodeBlock implements IIcfgCallTransition<IcfgLocation>
 	protected String mPrettyPrintedStatements;
 
-	Call(final int serialNumber, final BoogieIcfgLocation source, final BoogieIcfgLocation target,
+	public Call(final int serialNumber, final BoogieIcfgLocation source, final BoogieIcfgLocation target,
```
Why is it necessary to make these constructors public?

maul-esel replied:
ping

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/preferences/TAPreferences.java:
```diff
@@ -96,6 +96,16 @@ public final class TAPreferences {
+	public boolean minimizeAbstractionPerWorker;
+	public boolean mUseQuickCheckWorker;
+	public int mQuickCheckLoopBound;
+	public int mSearchLoopBound;
```
make these final, private, and follow the naming convention.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/preferences/TraceAbstractionPreferenceInitializer.java:
```diff
@@ -512,6 +512,29 @@ public enum CoinflipMode {
+	// Parallel Trace Abstraction
+	// ========================================================================
+	public static final String LABEL_PARALLEL_CEGAR_LOOP = "Use CEGAR loop for Parallel Trace Abstraction";
+	public static final boolean DEF_PARALLEL_CEGAR_LOOP = false;
```
minor: the `DEF_` constants are typically private

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/preferences/TraceAbstractionPreferenceInitializer.java:
```diff
@@ -724,7 +747,7 @@ private static UltimatePreferenceItemContainer getConcurrencySettings() {
 	}
 
-	private static UltimatePreferenceItemContainer getPORSettings() {
+	public static UltimatePreferenceItemContainer getPORSettings() {
```
why do you need to make this public?

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/tracehandling/StrategyFactory.java:
```diff
@@ -109,6 +110,22 @@ public class StrategyFactory<L extends IIcfgTransition<?>> {
+		mPathProgramCache = programCache;
+		mTransitionClazz = transitionClazz;
+	}
+
```
please avoid the code duplication between constructors

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/tracehandling/StrategyFactory.java:
```diff
@@ -154,18 +172,34 @@ public ITARefinementStrategy<L> constructStrategy(final IUltimateServiceProvider
-	 *
-	 *            Optionally, accompanied by the sequence of control configurations visited by the trace in the program
-	 *            that is being verified. This sequence is used to judge the quality of proofs ("perfect") and for
-	 *            assert order modulation.
```
Please do not remove this documentation.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/tracehandling/TraceAbstractionRefinementEngine.java:
```diff
@@ -74,6 +75,19 @@ public final class TraceAbstractionRefinementEngine<L extends IIcfgTransition<?>
+		mAfeStatistics = afEngine.getRefinementEngineStatistics();
+		generateProof();
+	}
+
```
Please avoid code duplication between the constructors

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/AbstractCegarLoop.java:
```diff
@@ -139,7 +142,7 @@ public abstract class AbstractCegarLoop<L extends IIcfgTransition<?>, A extends
 	 * Current Iteration of this CEGAR loop.
 	 */
-	private int mIteration;
+	protected int mIteration;
```
Is it necessary to make this protected? AFAIR there is a getter method anyway.

maul-esel replied:
ping

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/BasicCegarLoop.java:
```diff
@@ -140,23 +140,51 @@
+			return mShortString;
+		}
+
+	}
```
I think these additions are the result of merge issues.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/BasicCegarLoop.java:
```diff
@@ -599,20 +624,17 @@ private NondeterministicInterpolantAutomaton<L> constructInterpolantAutomatonFor
+			final NestedWordAutomaton<L, IPredicate> inputInterpolantAutomaton,
+			final IPredicateUnifier predicateUnifier, final IHoareTripleChecker htc,
+			final InterpolantAutomatonEnhancement enhanceMode) {
+		final boolean conservativeSuccessorCandidateSelection = enhanceMode == InterpolantAutomatonEnhancement.PREDICATE_ABSTRACTION_CONSERVATIVE;
```
There are many formatting-only changes here, making it hard to review. Please ensure you use the official formatting settings.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarLoopFactory.java:
```diff
@@ -221,16 +221,25 @@ private NwaCegarLoop<L> createFiniteAutomataCegarLoop(final IUltimateServiceProv
+			return new NwaCegarLoop<>(name, abstraction, root, csToolkit, predicateFactory, mPrefs, errorLocs,
+					proofProducer, services, mTransitionClazz, stateFactoryForRefinement);
+		default:
+			throw new AssertionError("Unknown Setting: " + mPrefs.getFloydHoareAutomataReuse());
```
Please do not add a `default` case here.

The previous version of the switch was exhaustive, and the exhaustiveness was checked by the Java compiler. This ensures that if in the future, new enum values are added, we get a compiler error and see immediately which code needs to be adapted. The `default` case undermines this.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarNwaWorkerThread.java:
```diff
@@ -0,0 +1,569 @@
+package de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction;
```
Remember to add license headers to all new files.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/NwaCegarLoop.java:
```diff
@@ -159,13 +159,19 @@ public NwaCegarLoop(final DebugIdentifier name, final INestedWordAutomaton<L, IP
+			final Set<IPredicate> possibleEndPoints) throws AutomataOperationCanceledException {
+		return new IsEmpty<>(new AutomataLibraryServices(mServices), abstraction, abstraction.getInitialStates(),
+				Collections.emptySet(), possibleEndPoints).getNestedRun();
+
```
Unless I missed a call, this method seems not to be called within this class. If only your subclass needs it, why not declare it there?

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/ParallelNwaCegarLoop.java:
Please do not declare multiple classes in the same file (except for nested classes).

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/PathProgramCache.java:
```diff
@@ -48,8 +48,8 @@
-	private final Map<Set<LETTER>, Integer> mKnownPathPrograms;
-	private final List<Integer> mTraceHashes;
+	private  Map<Set<LETTER>, Integer> mKnownPathPrograms;
+	private  List<Integer> mTraceHashes;
```
Why not keep these as `final`?

maul-esel commented on trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/RCFGBacktranslator.java:
```diff
@@ -242,11 +242,13 @@ private void addCodeBlock(final IIcfgTransition<IcfgLocation> cb, final IRelevan
 			throw new IllegalArgumentException();
 		}
+		
+		mTerm2Expression.getTypeSortTranslator().setCfgScriptFromWorker(rcfgProgramExecution.getOriginCfgScript());
```
Why is this call necessary here?

maul-esel replied:
(see my comment in `TypeSortTranslator`)

maul-esel reviewed (commented):
I left a number of comments. I focused mainly on changes to existing code (I do not have enough overview to review the newly-added parallel CEGAR itself).

MaxBarth95 committed:
reafactoring searches

MaxBarth95 committed:
addressed some merge comments

MaxBarth95 committed:
code cleanup for merge request

MaxBarth95 committed:
code cleanup searches, removed noLoopMode flag from isempyheuristic since it was no longer used

MaxBarth95 committed:
readded lock

MaxBarth95 committed:
cleaned up IsEmptyParallel

MaxBarth95 committed:
removed unncesa code, cleanup

MaxBarth95 committed:
added comment

MaxBarth95 committed:
cleanup

MaxBarth95 committed:
cleanup basicCegarLoop

MaxBarth95 committed:
cleanup CEGARloopfactory

MaxBarth95 committed:
fix build by adding exception

MaxBarth95 committed:
cleanup NWA

MaxBarth95 committed:
cleanup TAPreferences

MaxBarth95 committed:
cleanup perference initializer

MaxBarth95 committed:
cleanup perference initializer

MaxBarth95 committed:
cleanup perference initializer

MaxBarth95 committed:
cleaned StrategyFactory

MaxBarth95 committed:
and the rest of the strategy stuff

MaxBarth95 committed:
and the rest of the strategy stuff

MaxBarth95 committed:
and the rest of the strategy stuff

MaxBarth95 committed:
clean pathprogram cache

MaxBarth95 committed:
use Excutionservice again, add licence hadders

MaxBarth95 committed:
Fixed NoInterpolationWorker, cleaned up the rest

MaxBarth95 commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TransferBetweenMainAndWorker.java:
```diff
@@ -0,0 +1,547 @@
+					}
+				}
+			}
+			// TODO transfer summaries, so far they are ignored because i dont know how to detect a summary in the input
```
WIP

MaxBarth95 committed:
bugfix

MaxBarth95 committed:
bugfix

MaxBarth95 committed:
removing noInterpolation worker

MaxBarth95 committed:
cleanup nointerpolation worker

MaxBarth95 committed:
fix

MaxBarth95 committed:
clean

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmpty.java:
```diff
@@ -345,6 +382,7 @@ private DoubleDecker<STATE> dequeue() {
 		// If available, take a state pair that has been discovered by taking an internal or a return transition or a
 		// summary. If not, take a state pair that has been discovered by taking a call transition.
 		case DFS -> dequeueGivenQueues(mQueue, mQueueCall);
+		default -> throw new IllegalArgumentException("Unexpected value: " + mStrategy);
```
Please avoid adding a `default` case to this previously-exhaustive switch.

Ideally, `IsEmpty` should not need special handling for its subclass `IsEmptyParallel`. Do you really need the additional enum value? (it's also a bit strange since it couples a search mode (BFS) with the fact that the search is parallelized -- can they be split?)

If there is no way to avoid having the additional enum value, use `case PARALLEL ->` here.

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyHeuristic.java:
```diff
@@ -116,7 +118,25 @@ public IsEmptyHeuristic(final AutomataLibraryServices services,
+
+	/**
+	 * Default constructor. Here we search a run from the initial states of the automaton to the final states of the
+	 * automaton and use the zero heuristic.
```
This documentation seems incorrect (the caller actually passes a heuristic) and does not mention the meaning of the `wayPoints` parameter.

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyHeuristic.java:
```diff
@@ -128,15 +148,15 @@ public IsEmptyHeuristic(final AutomataLibraryServices services, final INestedWor
 			throws AutomataOperationCanceledException {
-		this(services, (INwaOutgoingLetterAndTransitionProvider<LETTER, STATE>) operand, startStates,
-				funIsForbiddenState, funIsGoalState, heuristic);
+		this(services, operand, startStates, funIsForbiddenState, funIsGoalState, heuristic, new ArrayList<>());
```
minor: use `List.of()` or `Collections.emptyList()` to make it clear the list is unmodifiable

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyParallel.java:
```diff
@@ -0,0 +1,711 @@
+ * counterexample. Then we call BFS in this arbitrary state to find a path (suffix) to a goal state. At the end both
+ * paths are merged.
+ *
+ * This class uses recursion. It recursively explores the successor of a state. TODO: non recursive
```
Are you planning to address this TODO?

MaxBarth95 replied:
Not in the near future

maul-esel commented on trunk/source/Library-Automata/src/de/uni_freiburg/informatik/ultimate/automata/nestedword/operations/IsEmptyParallel.java:
Does the constructor you are referencing here actually exist?

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/preferences/TraceAbstractionPreferenceInitializer.java:
```diff
@@ -512,6 +512,24 @@ public enum CoinflipMode {
+	private static final boolean DEF_MINIMIZE_ABSTRACTION_PER_WORKER = true;
+	// Parallel CEGAR counterexample search stragies
+	// ========================================================================
+	public static final String LABEL_PARALLELSEARCH_ACTIVE_CEX_ONLY = "Consider only active in Search Strategy";
```
Could you add a more detailed description for this setting?

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/preferences/TraceAbstractionPreferenceInitializer.java:
```diff
@@ -512,6 +512,24 @@ public enum CoinflipMode {
+	// ========================================================================
+	public static final String LABEL_PARALLELSEARCH_ACTIVE_CEX_ONLY = "Consider only active in Search Strategy";
+	private static final boolean DEF_PARALLELSEARCH_ACTIVE_CEX_ONLY = false;
+	public static final String LABEL_SEARCH_LOOP_BOUND = "search loop bound";
```
Could you add a more detailed description for this setting?

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/BasicCegarLoop.java:
```diff
@@ -156,6 +156,8 @@ public abstract class BasicCegarLoop<L extends IIcfgTransition<?>, A extends IAu
 	protected boolean mFallbackToFpIfInterprocedural = false;
 	protected IRefinementEngineResult<L, NestedWordAutomaton<L, IPredicate>> mRefinementResult;
+	protected InterpolationTechnique mInterpolationTechnique;
+	protected Class<L> mTransitionClazz;
```
These two fields seem unused within this class. If they are only useful for a subclass, could you not declare them in the subclass directly?

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarLoopFactory.java:
```diff
@@ -221,16 +221,24 @@ private NwaCegarLoop<L> createFiniteAutomataCegarLoop(final IUltimateServiceProv
+			return new NwaCegarLoop<>(name, abstraction, root, csToolkit, predicateFactory, mPrefs, errorLocs,
+					proofProducer, services, mTransitionClazz, stateFactoryForRefinement);
+		}
+		throw new AssertionError("Unknown Setting: " + mPrefs.getFloydHoareAutomataReuse());
```
As for other switches, please rewrite this such that it is an exhaustive switch again.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarNwaWorkerThread.java:
```diff
@@ -0,0 +1,578 @@
+				mBlockingQueueForResults.put(mThreadResult);
+			} catch (final InterruptedException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
```
Please replace this auto-generated block.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/CegarNwaWorkerThread.java:
```diff
@@ -0,0 +1,578 @@
+
+	/*
+	 * The worker creates for each counterexample a new strategy module. We can switch between different
+	 * RefinementStrateies here. For example we can use @RefinementStrategy.ACCELERATED_TRACE_CHECK if we have seen the
```
nitpick: the javadoc syntax is `{@link RefinementStrategy.ACCELERATED_TRACE_CHECK}`, not `@RefinementStrategy.ACCELERATED_TRACE_CHECK`

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/PathProgramCache.java:
```diff
@@ -53,8 +53,8 @@ public class PathProgramCache<LETTER> {
-		mKnownPathPrograms = new HashMap<>();
-		mTraceHashes = new ArrayList<>();
+		mKnownPathPrograms = new ConcurrentHashMap<>();
+		mTraceHashes = Collections.synchronizedList(new ArrayList<>());
```
This seems intended to make this class thread-safe, which it wasn't intended to be by the original author.

I don't have a particular objection to this change (though it might have a bit of a performance impact, it should not be critical).

But it leads me to wonder which other classes are out there with similar problems, that will only be detected by hard-to-reproduce concurrency bugs. Do you have somewhere a documentation of which mutable objects are shared between threads, and how thread safety is ensured for them?

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TransferBetweenMainAndWorker.java:
```diff
@@ -0,0 +1,547 @@
+		}
+	}
+
+	private UnmodifiableTransFormula transferTransformula(final UnmodifiableTransFormula inTF,
```
We already have methods `TransferrerWithVariableCache::transferTransFormula` resp. `TransFormulaBuilder::transferTransformula`. Is your implementation doing something fundamentally different, or could you reuse those?

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TransferBetweenMainAndWorker.java:
```diff
@@ -0,0 +1,547 @@
+				constructNewProc2Globals(mMainCsToolkit.getModifiableGlobalsTable().getProcToGlobals(), mVarTransfer);
+		final ModifiableGlobalsTable modifiableGlobalsTable = new ModifiableGlobalsTable(proc2globals);
+		final IIcfgSymbolTable symbolTable =
+				constructNewSymbolTable(mMainCsToolkit.getSymbolTable(), mMainCsToolkit.getProcedures(), mVarTransfer);
```
`TransferrerWithVariableCache::transferSymbolTable` could help here.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/TransferBetweenMainAndWorker.java:
```diff
@@ -0,0 +1,547 @@
+
+}
+
+class VariableTransferrer {
```
Please don't add multiple top-level classes in the same file.

maul-esel commented on trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/WorkerThreadResult.java:
```diff
@@ -0,0 +1,99 @@
+package de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction;
```
Please add a license header

maul-esel reviewed (commented):
Thanks for continuing the work on this 👍 

I've gone through the changes again and left some comments. I very much appreciate the much more decoupled approach compared to the first implementation 👍

MaxBarth95 committed:
Bug fixe in IsEmptyParallel for recursion. Make sure we dont use ParallelCegarLoop in case of a concurrent program.

MaxBarth95 committed:
adressing review comments

MaxBarth95 committed:
fix documentation

MaxBarth95 committed:
use emptylist

MaxBarth95 committed:
potential bugfix need to evaluate, adressed comments

MaxBarth95 committed:
moved sort transfer from TypeSortTranslator to Term2Expression

MaxBarth95 committed:
Merge remote-tracking branch 'origin/dev' into merge/mb/parallel-trace-abstraction

MaxBarth95 committed:
moved to a new file

MaxBarth95 committed:
adresses review

MaxBarth95 committed:
added description to new settings

MaxBarth95 committed:
make private final

MaxBarth95 committed:
minor change

MaxBarth95 committed:
copying the variable manager content to preven the varning of unknown TV being triggered

MaxBarth95 committed:
missing

MaxBarth95 committed:
wip finalize merge

MaxBarth95 committed:
Merge remote-tracking branch 'origin/dev' into merge/mb/parallel-trace-abstraction

MaxBarth95 committed:
Merge remote-tracking branch 'origin/dev' into merge/mb/parallel-trace-abstraction

MaxBarth95 committed:
new comment

MaxBarth95 committed:
Merge remote-tracking branch 'origin/dev' into merge/mb/parallel-trace-abstraction

MaxBarth95 committed:
final points of MR adressed

MaxBarth95 committed:
fix build

MaxBarth95 committed:
Merge remote-tracking branch 'origin/dev' into merge/mb/parallel-trace-abstraction

MaxBarth95 committed:
changed comment
